evolved steps towards devicetree-based device definitions and dependency representations

pabigot opened the pull request:
https://github.com/zephyrproject-rtos/zephyr/pull/26616#issuecomment-672012086 has the most recent status summary.  **Please look at that before reviewing** so you can focus on what's supposed to be in this PR separate from the supporting capabilities.

That summary is updated every time this PR is pushed, most recently 2020-08-21T10:26-05:00.

This supersedes #26388 which is similar but was a trial for whether the handles could be identified solely from devictree.  They can't (#26575), so this shows current status of an approach that post-processes the prebuilt ELF file and will generate correct and complete dependency data.

The solution here extends the `DEVICE_INIT()` macro infrastructure to take the declaration names and other information from a devicetree node.  For devices defined in this way dependency information from the devicetree and from external sources are stored in an array referenced from the driver structure.  This array uses the ordinals from the devicetree, and is placed in a linker section that will be discarded in the final link.

A new script `gen_handles.py` reads the prebuilt binary and uses the device handle array data to associate each device with the devicetree node associated with it.  The dependency graph of the whole devicetree is then reduced to a dependency graph involving only devices present in the application.  Ordinals for these devices are assigned based on their offset in the Zephyr device section, and new handle values are generated that replace the ones used in the pre-build link.

Issues to be addressed:
* Dependency information is available only when the drivers use devicetree nodes to define the device structure.  All drivers will have to be updated to do this.
* Many devices will not have any recorded dependencies.  Each of those currently costs 2 bytes for the device-specific handle array that holds the empty sequence.  Those can be replaced by aliases to a common empty sequence array.
* Although non-device handles can be added there is currently no mechanism to define them for reference from other devices and init functions.

The tooling-supplied ordinal is not used in the application.  The rework would enable:

    struct device *dev = &DEVICE_DT_GET(nodeid);

as a mechanism to access the device structure, if the underlying names were made global.

pabigot committed:
device: provide API that uses the devicetree node as identifier

The dev_name is the canonical DT node identifier encoding the
devicetree path; the drv_name is the label of the corresponding node.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP drivers: timer: demonstrate way of associating handles with system devices

The system clock is a device that has no identifiable devicetree node,
so is not assigned an ordinal that can be used as its handle.  Use an
externally-specified ordinal for this device.

pabigot committed:
DNM hello: playground for device dependencies

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: redefine devicetree device definition using a more generic API

For devices defined using devicetree nodes we need to pass the
devicetree node, and to infer names for the device and the driver from
it.  Devices defined in the classical way don't need this.  Introduce
another level of private abstraction that accepts the information from
either source, falling back to the old parameters when the provided
node is invalid.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: inject device dependency information from devicetree nodes

Generate arrays of dependency information for each device.  If a
device definition is being constructed from devicetree these come from
the devicetree dependency information.  Additional dependencies may be
passed through using the DT_ macros.

Define flag values for device handles so we can partition the
dependency array into distinct sets, which include things it requires,
things it supports (may not be needed), and child nodes (not
implemented, may not be needed).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
device: add post-process of elf file to manage device handles

Following the idiom used for system calls, add script support to read
the initial application binary to identify which devices are defined,
and to use their offset in the device array as their unique handle
rather than the externally-defined ordinal from devicetree.  The
device dependency arrays are updated to use these handles.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

cfriedt commented:
Good work. I recently did something like this for a bare metal module loader at $DAYJOB and took the same approach, reducing the problem to a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting).

pabigot committed:
DNM samples: spi_flash: hack for testing

pabigot commented:
Still not ready for real review, but this is basically functional now.  @mbolivar-nordic please look; the commits I've adopted cover the current need for devicetree API for ordinals.  That the original devicetree ordinals are ephemeral (replaced by the index of the device in the device array) suggests that exposing generic API to access them in C is probably not useful.  So we're much closer now to closing on #26062.

@galak if you feel it worthwhile we could discuss this in dev-review tomorrow.  It's going to be blocked until the changes to support defining devices with devicetree nodes go in, which probably conflict with #24873 or other work @tbursztyka is doing, so I don't think there's a great rush.

cfriedt commented:
Definitely could use deterministic device intialization  right about now.

mbolivar-nordic commented:
>  @mbolivar-nordic please look; the commits I've adopted cover the current need for devicetree API for ordinals.

I haven't forgotten this and will get to it as soon as I can.

mbolivar-nordic commented on scripts/gen_handles.py:
```diff
@@ -0,0 +1,290 @@
+        format += "%uh" % (len(data) / 2)
+        return struct.unpack(format, data)
+
+DevicePtrs = collections.namedtuple('DevicePtrs', [
```
A comment in the `struct device` definition mentioning this file and that it needs to be kept in sync would not go amiss.

pabigot replied:
Actually these need to use material provided by `GEN_OFFSET_SYM()`.

mbolivar-nordic reviewed (commented):
> @mbolivar-nordic please look [...] That the original devicetree ordinals are ephemeral (replaced by the index of the device in the device array) suggests that exposing generic API to access them in C is probably not useful.

I agree and suggest we hide ordinal access APIs in the `Z_` namespace.

The approach still seems reasonable after digging into the details. I agree with your list of main items that still really need to be addressed.

pabigot commented on scripts/dts/edtlib.py:
```diff
@@ -516,6 +522,7 @@ def _init_luts(self):
         # Initialize node lookup tables (LUTs).
 
         self.label2node = OrderedDict()
+        self.depord2node = OrderedDict()
```
OK, I've annotated the commit message with that commitment, will leave as-is.

pabigot commented on include/device.h:
```diff
@@ -124,33 +124,25 @@ extern "C" {
-	};								\
-	Z_INIT_ENTRY_DEFINE(_CONCAT(__device_, dev_name), init_fn,	\
-			    (&_CONCAT(__device_, dev_name)), level, prio)
+	Z_DEVICE_DEFINE(_, dev_name, drv_name, init_fn, pm_control_fn,	\
```
Commit for this feature added for next push.

pabigot commented on include/device.h:
```diff
@@ -128,26 +155,40 @@ extern "C" {
 			data, cfg_info, level, prio, api)
```
I'm going to say intentionally commited.  For this whole initiative to work we need to use the `DEVICE_DT_INIT()` macros everywhere that information can come from devicetree.

It's yet unclear whether we will be trying to capture non-device dependencies at build time, or whether they're going to be registered at runtime through the "Resource Manager".  Until that's resolved we can't tell whether we need non-device handles at build time.

pabigot commented on include/device.h:
```diff
@@ -225,6 +265,49 @@ struct device {
+}
+
+static inline device_handle_t
+device_get_parent_ord(const struct device *dp)
```
One is; the others just needed an update.

pabigot commented:
Currently this draft PR has the following content:

* #27443 which extends the on-off service in support of...
> d1dfee381d sys: onoff: use updated API terminology
> 0c59be0d73 sys: onoff: support clients using synchronous transitions

* #27360 which adds regulator devices which provide another source of meaningful dependency in devicetree.
> bbb521e6bb devicetree: add infrastructure for regulator devices
> ae865fb766 drivers: regulator: add GPIO-controlled regulator driver
> d61f8e2a80 devicetree: i2c-device: add power-supply property
> 9e98876ac4 boards: thingy52_nrf52832: add nodes for on-board voltage regulators

* Pieces from #26062 by @mbolivar-nordic providing devicetree macro support to get dependency information from devicetree
> f2665de9a6 devicetree: add first round of dependency ordinal info
> 051e9b0339 doc: document devicetree.h dependency ordinal APIs

* Additional commits the complement or extend what's provided in #26062
> 971e3efb59 DNM: add FOR_EACH_NONEMPTY_TERM
> a1cdb117b9 REWORK edtlib: add support for finding node from dependency ordinal
> 2e8b323222 sys: util: add identity macro
> 2cfae1b56d devicetree: add a name for an invalid node identifier

* Commits that rework the device initialization macros so that devices can be defined using devicetree node data.  These will conflict with other device restructuring work in progress
> 6183da35ec device: provide API that uses the devicetree node as identifier
> a5f68084cf device: redefine devicetree device definition using a more generic API

* Partial conversion of existing drivers to use the new `DEVICE_DT_` macros to define devices, so that they can participate in dependency relationships
> 9fe52be8c4 WIP drivers/clock: update to DT device API
> 5cc9201dc3 WIP drivers/flash: update to DT device API
> b7d41103ac WIP drivers/gpio: update to DT device API
> c7d2a833db WIP drivers/i2c: update to DT device API
> f25765733b WIP drivers/regulator: update to DT device API
> 7e1dcdd9ab WIP drivers/sensor: update to DT device API
> 0c61635c83 WIP drivers/serial: update to DT device API
> f5094ce578 WIP drivers/spi: update to DT device API

* Extending the device structure to represent dependency information using handles and information from the devicetree nodes
> 050caf68b8 WIP device: inject device dependency information from devicetree nodes
> f9f003bf9b WIP timer: demonstrate way of associating handles with system devices

* Translating the devicetree ordinals to handles that directly correspond to device objects
> 280f859f56 REWORK scripts/gen_handles: add post-processing to fix up handle references

* Hacked examples that display the dependency information
85df75d05b DNM hello: playground for device dependencies
3a4a47c717 DNM samples: spi_flash: hack for testing

Dependency information for arbitrary programs can be obtained using the gen_handles script, e.g.:

```
python3 scripts/gen_handles.py -v -o g.c \
  -k samples/sensor/ccs811/build/zephyr/zephyr_prebuilt.elf
```

That sample, when built for thingy52_nrf52832, has the deepest dependency set currently available.

pabigot committed:
drivers: clock_control: convert nrf to dt device defines

Use the clock devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: flash: convert spi-nor drivers to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio: convert nrf driver to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: i2c: convert nrf drivers to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: regulator: convert to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: sensor: ccs811: convert to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: serial: convert nrf drivers to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: spi: convert nrf drivers to dt device defines

Use the devicetree node as the source of object name and other
information used when defining the device structure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
